mender-auth/cli/keystore.cpp

```diff
@@ -0,0 +1,105 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include <mender-auth/cli/keystore.hpp>
+
+#include <common/log.hpp>
+#include <common/crypto.hpp>
+
+
+namespace mender {
+namespace auth {
+namespace cli {
+
+using namespace std;
+
+namespace log = mender::common::log;
+
+namespace crypto = mender::common::crypto;
+
+class KeyStoreErrorCategoryClass : public std::error_category {
+public:
+	const char *name() const noexcept override;
+	string message(int code) const override;
+};
+
+enum ErrorCode {
+	NoError = 0,
+	NoKeysError,
+	StaticKeyError,
+};
+
+const KeyStoreErrorCategoryClass KeyStoreErrorCategory;
+
+const char *KeyStoreErrorCategoryClass::name() const noexcept {
+	return "KeyStoreErrorCategory";
+}
+
+string KeyStoreErrorCategoryClass::message(int code) const {
+	switch (code) {
+	case NoError:
+		return "Success";
+	case NoKeysError:
+		return "No keys found";
+	case StaticKeyError:
+		return "Cannot replace static key";
+	}
+	// Don't use "default" case. This should generate a warning if we ever add any enums. But
+	// still assert here for safety.
+	assert(false);
+	return "Unknown";
+}
+
+const error::Error KeyStoreNoKeysError =
+	error::Error(error_condition(NoKeysError, KeyStoreErrorCategory), "no keys");
+
+const error::Error KeyStoreStaticKeyError = error::Error(
+	error_condition(StaticKeyError, KeyStoreErrorCategory), "cannot replace static key");
+
+error::Error MenderKeyStore::Load() {
+	auto exp_key = crypto::PrivateKey::LoadFromPEM(key_name_, passphrase_);
+	if (!exp_key) {
+		return KeyStoreNoKeysError;
+	}
+	log::Debug("Successfully loaded private key from " + key_name_);
+	key_ = move(exp_key.value());
+
+	return error::NoError;
+}
+
+error::Error MenderKeyStore::Save() {
+	if (!key_) {
+		return KeyStoreNoKeysError;
+	}
+
+	return key_.get()->SaveToPEM(key_name_);
+}
+
+error::Error MenderKeyStore::Generate() {
+	if (static_key_) {
+		return KeyStoreStaticKeyError;
+	}
+
+	auto exp_key = crypto::PrivateKey::Generate(3072);
```

Comment:
const at the top of the file?

Reply:
What is the reasoning? To reuse in case of re-generation? I don't really see the benefit :see_no_evil:

Reply:
It struck me cause I was once surprised that we appareantly enforce the key-length in our code, and it's just easier to find that way _I think_. But no hard requirement here, YMMV as they say

Reply:
Oh man, you mean the magic number `3072`! Of course.

I miss-interpreted your comment earlier and I thought you meant that `exp_key` object could be allocated and constant somewhere :weary:

Reply:
tbh, I think this should even be a configurable parameter in our build :angel: